File: src/main/java/edu/cornell/cs/apl/viaduct/security/solver/ConstantJoinVariableTerm.java
package edu.cornell.cs.apl.viaduct.security.solver;

import com.google.auto.value.AutoValue;
import edu.cornell.cs.apl.viaduct.util.HeytingAlgebra;
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge;

/** Join of a constant element and a variable. */
@AutoValue
public abstract class ConstantJoinVariableTerm<A extends HeytingAlgebra<A>>
    implements RightHandTerm<A> {
  public static <A extends HeytingAlgebra<A>> ConstantJoinVariableTerm<A> create(
      A lhs, VariableTerm<A> rhs) {
    return new AutoValue_ConstantJoinVariableTerm<>(lhs, rhs);
  }

  protected abstract A getLhs();

  protected abstract VariableTerm<A> getRhs();

  @Override
  public final ConstraintValue<A> getNode() {
    return getRhs();
  }

  @Override
  public final DataFlowEdge<A> getOutEdge() {
    return new PseudocomplementEdge<>(getLhs());
  }
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/solver/ConstantJoinVariableTerm.kt
package io.github.apl_cornell.viaduct.algebra.solver

import io.github.apl_cornell.viaduct.algebra.HeytingAlgebra
import io.github.apl_cornell.viaduct.util.dataflow.DataFlowEdge

/** Join of a constant element and a variable. */
internal data class ConstantJoinVariableTerm<A : HeytingAlgebra<A>>(val lhs: A, val rhs: VariableTerm<A>) :
    RightHandTerm<A> {
    override fun getValue(solution: ConstraintSolution<A>): A {
        return lhs.join(rhs.getValue(solution))
    }

    override val node: AtomicTerm<A>
        get() = rhs

    override val outEdge: DataFlowEdge<A>
        get() = JoinEdge(lhs)
}
